libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,206 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/FunctionCallGraph.h>
+#include <boost/range/adaptor/reversed.hpp>
+
+using namespace std;
+using namespace solidity::frontend;
+
+FunctionCallGraphBuilder::FunctionCallGraphBuilder(ContractDefinition const& _contract):
+	m_contract(&_contract),
+	m_graph(std::make_unique<ContractCallGraph>(_contract))
+{
+	// Create graph for constructor, state vars, etc
+	m_currentNode = SpecialNode::EntryCreation;
+	m_currentDispatch = SpecialNode::InternalCreationDispatch;
+
+	for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts | boost::adaptors::reversed)
+	{
+		for (auto const* stateVar: contract->stateVariables())
+			stateVar->accept(*this);
+
+		for (auto arg: contract->baseContracts())
+			arg->accept(*this);
+
+		if (contract->constructor())
+		{
+			add(*m_currentNode, contract->constructor());
+			contract->constructor()->accept(*this);
+			m_currentNode = contract->constructor();
+		}
+	}
+
+	m_currentNode.reset();
+	m_currentDispatch = SpecialNode::InternalDispatch;
+
+	// Create graph for all publicly reachable functions
+	for (auto& [hash, functionType]: _contract.interfaceFunctionList())
+	{
+		(void)hash;
+		if (auto const* funcDef = dynamic_cast<FunctionDefinition const*>(&functionType->declaration()))
+			if (!m_graph->edges.count(funcDef))
+				visitCallable(funcDef);
+
+		// Add all external functions to the RuntimeDispatch
+		add(SpecialNode::Entry, &functionType->declaration());
+	}
+
+	// Add all InternalCreationDispatch calls to the RuntimeDispatch as well
+	add(SpecialNode::InternalDispatch, SpecialNode::InternalCreationDispatch);
+
+	if (_contract.fallbackFunction())
+		add(SpecialNode::Entry, _contract.fallbackFunction());
+
+	if (_contract.receiveFunction())
+		add(SpecialNode::Entry, _contract.receiveFunction());
+}
+
+bool FunctionCallGraphBuilder::CompareByID::operator()(Node const& _lhs, Node const& _rhs) const
+{
+	if (_lhs.index() != _rhs.index())
+		return _lhs.index() < _rhs.index();
+
+	if (std::holds_alternative<SpecialNode>(_lhs))
+		return std::get<SpecialNode>(_lhs) < std::get<SpecialNode>(_rhs);
+	return std::get<ASTNode const*>(_lhs)->id() < std::get<ASTNode const*>(_rhs)->id();
```

Comment:
```suggestion
	if (holds_alternative<SpecialNode>(_lhs))
		return get<SpecialNode>(_lhs) < get<SpecialNode>(_rhs);
	return get<ASTNode const*>(_lhs)->id() < get<ASTNode const*>(_rhs)->id();
```

There are also a few more instances below.

Reply:
Done.